### ISSUE TITLE
use Microsoft tools to build lisp kernel [$700]

### ISSUE THREAD

**Issue by xrme** [Clozure/ccl#59]
I think it would be worthwhile to make the lisp kernel buildable with Microsoft's toolchain (either or both of the command-line tools or with Visual Studio).

My understanding is that this would let us to Microsoft's debugging tools (such as windbg), which may be better (in some sense) than using gdb.

One challenge is going to be dealing with the various assembly language files.  There's an assembler called yasm (http://yasm.tortall.net) which should be able to support the AT&T style syntax that we use.

**Comment by xrme** [comment]
To elaborate on this a little bit more:

On Windows, we build the lisp kernel with Cygwin or msys2.  Please see the comments at the top of the Makefile in the appropriate lisp kernel build directory (e.g., https://github.com/Clozure/ccl/blob/master/lisp-kernel/win64/Makefile).

As mentioned already, the main task to building with some other toolchain is going to be building the assembly-language files.  The `.s` files are processed with m4, since system assemblers have historically been pretty weak in terms of supporting macro facilities.

If we can come up with some way to do this in the context of building with Microsoft's toolchain, this might also make it easier to build the lisp kernel with Xcode, as the linked issue mentions.

**Comment by ailisp** [comment]
I got some time this weekend, will take a look. And if this comment also post on bountysource will show bountysource and github linking is suceessful.

**Comment by xrme** [comment]
See @ailisp's PR #139

**Comment by denim2x** [comment]
@xrme [Here](https://github.com/denim2x/ccl/wiki/Lisp-kernel:-compatibility-with-WinDbg) are some possible approaches for solving this; please let me know what you think

**Comment by xrme** [comment]
@denim2x I think using clang would be fine.  There is no requirement to use Microsoft's compiler.  Just as a reminder, the idea behind all of this is to be able to use windbg and possibly other Microsoft debugging tools instead of using gdb running under Cygwin.

The trouble has always been the assembly-language files.  The first challenge is running them through m4 as part of the build process.  I know there are Windows versions of GNU m4 out there.

Is there an assembler that goes along with the clang/LLVM toolchain?  If so, and if it supports the AT&T syntax, then that's probably the way to go.

**Comment by denim2x** [comment]
@xrme According to the [docs](https://clang.llvm.org/compatibility.html#inline-asm), the 'LLVM integrated assembler is extremely compatible with GAS'; however, there are a few places where LLVM imposes *stricter* rules, for instance in the case of 'ambiguous X86 instructions that don't have suffixes' - to illustrate:
```asm
add $4, (%rax)
```
In this snippet 'there is nothing that specifies the size of the instruction' and is thus rejected by Clang with this error message:
```asm
# ambiguous instructions require an explicit suffix (could be 'addb', 'addw', 'addl', or 'addq')
add $4, (%rax)
^
```
This is resolved by adding 'an explicit suffix to the instruction', thus making the code 'compatible with both GCC and Clang'.
To summarize, Clang requires all assembly instructions to have explicit suffixes, meaning all assembly code has to be inspected for missing suffixes and fixed accordingly. If this is the desired approach I'm willing to participate in this endeavor and hopefully have a working solution soon.

**Comment by xrme** [comment]
I think we already (mostly) use explicit suffixes, and it's certainly no problem to add them if some assembler stricter than `gas` requires them.

So, if you or anyone would like to move ahead with the approach of using the llvm toolchain to produce a Windows binary (that includes the debugging info that windbg can use), then I'd say please go right ahead.

**Comment by snunez1** [comment]
What about 'integration' with the rest of the MS ecosystem? For example being easily able to wrap the win32/64 API? Is this possible with Clang?

A major reason for funding this is to be able to get at the MS APIs easily.

**Comment by denim2x** [comment]
[Here](https://github.com/denim2x/ccl/tree/clang) is my preliminary work on Clang support; there are a few things left to be done

**Comment by xrme** [comment]
Building the lisp kernel with Microsoft-targeted tools is useful, but it's really not directly related to access to the Windows API.  We can already call win32 API functions via the FFI.

**Comment by denim2x** [comment]
okay; this is actually an attempt at obtaining Windows debug info via Clang with the MSVC backend

**Comment by snunez1** [comment]
That sounds like a worthy goal. How is it going?

**Comment by snunez1** [comment]
Is there a pull request for this? I'd like to give compiling with MSVC a try...

**Comment by xrme** [comment]
won't fix (at least for now)

**Comment by shawnl** [comment]
When I looked into this is was quite difficult and I gave up after a few hours working at it.
